File: references/openguardrails-main/openguardrails-main/integrations/agent/dsh/src/wire.ts
/**
 * The OGR v0.8 wire, hand-rolled.
 *
 * There is no SDK layer — the Runtime API is the integration surface
 * (specification/runtime-api.md), and since v0.8 an integration is ONE
 * decision endpoint: `/v1/evaluate` while holding an action (evaluate records
 * what it judges, so it is the observation channel too — `/v1/ingest` is
 * gone), plus the transport-level `/v1/heartbeat` that carries the
 * integration build id and the degraded-mode counters. This module is those
 * calls plus the wire types this plugin reads and writes, nothing else: no
 * signing, no batching machinery, no client-side decomposition.
 *
 * The base URL is joined with the canonical `/v1/...` paths exactly as the
 * binding requires — a deployment-specific prefix belongs IN the configured
 * base URL (`https://host/api/public/ogr`), never hard-coded here.
 */

/**
 * One v0.8 GuardEvent as this plugin sends it (snake_case, flat).
 *
 * Every field is REQUIRED — v0.8 removed every knob a producer could choose
 * to skip. What a runtime can derive is not on the wire at all (session,
 * turn, step numbering, timestamps, protocol versioning); what only the
 * producer can know is mandatory, with the empty string as the explicit
 * "I have nothing to assert" on the identity four-tuple. `step_id` is the
 * one coordinate kept: concurrency makes pairing a call's two halves
 * underivable, so the producer mints a fresh random id per model call and
 * puts the SAME value on both events.
 */
/**
 * WHO REPORTED IT — `"name/version"`, on every event AND on the heartbeat.
 * One constant so the two can never name different builds.
 */
export const INTEGRATION = "ogr-dsh/0.3.0"

export interface WireEvent {
  kind: "step/request" | "step/response"
  step_id: string
  agent_id: string
  agent_type: string
  agent_workspace: string
  /**
   * ⚠️ There is no `agent_owner` (removed 2026-08-17). Who is ACCOUNTABLE for an
   * agent is not something a producer can assert — it is a link to a console
   * account an administrator assigns, and a read permission cannot rest on a
   * claim the caller makes about itself.
   */
  agent_user: string
  llm_protocol: "openai.chat" | "openai.responses" | "anthropic.messages" | "canonical"
  payload: Record<string, unknown>
  /**
   * WHO REPORTED IT — `"name/version"`. The one OPTIONAL field on the v0.8
   * wire, stamped by {@link OgrClient.evaluate} so it cannot be forgotten at a
   * construction site.
   *
   * ⚠️ It rode the heartbeat ALONE until 2026-08-17, and that could not answer
   * "which build produced this traffic": a runtime keys its liveness record on
   * the integration NAME (it must, so a rollout updates its row instead of
   * minting a second and reporting the old build as dark), so every replica
   * overwrites the others' version. On the event the build travels with the
   * traffic it produced and nothing can overwrite it.
   */
  integration?: string
}

/** One v0.8 finding — what was found, where, and what it contributed. */
export interface WireFinding {
  category: string
  severity?: "low" | "medium" | "high" | "critical"
  action?: "flag" | "redact" | "block"
  path?: string
  start?: number
  end?: number
  score?: number
  detector?: string
  fp?: string
  whitelisted?: boolean
  subject?: string
}

/** The v0.8 Verdict: two decisions, findings, spans, and the coverage truth. */
export interface WireVerdict {
  event_id: string
  provider: string
  decision: "allow" | "block"
  latency_ms?: number
  findings?: WireFinding[]
  modifications?: { spans?: Array<{ path: string; start: number; end: number; replacement: string }> }
  /**
   * Payload paths this verdict could NOT judge. Absent/empty asserts every
   * routed text was judged; under `fail_mode: closed` a non-empty value is
   * "could not look", which is not "found nothing".
   */
  unjudged?: string[]
}

/**
 * One `/v1/heartbeat` body. NOT a GuardEvent — transport-level liveness so
 * the runtime can tell "agent idle" from "integration went dark". The
 * integration build id lives here (it left the event in v0.8), and the
 * counters are how an outage-induced observability gap stays visible: v0.8
 * has no replay channel, so a raised `evaluate_errors` is the record that
 * steps went unjudged.
 */
export interface WireHeartbeat {
  integration: string
  agent_id?: string
  interval_s?: number
  counters?: Record<string, number>
}

/** Log sink; the plugin passes dsh's own logger so wire noise stays in the harness log. */
export interface WireLog {
  info(message: string): void
  warn(message: string): void
}

/** One resolved runtime connection: where, and as whom. */
export interface RuntimeSource {
  url: string
  apiKey: string
}

/**
 * The evaluate client. The source is a THUNK, re-read on every call, so a
 * connection configured later — an API key pasted into the dsh Settings
 * form — takes effect without a restart.
 */
export class OgrClient {
  constructor(
    private readonly log: WireLog,
    private readonly source: () => RuntimeSource | null,
    private readonly timeoutMs: () => number,
  ) {}

  /** Whether a runtime is configured RIGHT NOW (the source is live). */
  get enabled(): boolean {
    return this.source() !== null
  }

  private async post(path: string, body: unknown): Promise<Response | null> {
    const src = this.source()
    if (!src) return null
    const base = src.url.endsWith("/") ? src.url.slice(0, -1) : src.url
    const controller = new AbortController()
    const timer = setTimeout(() => controller.abort(), this.timeoutMs())
    try {
      return await fetch(`${base}${path}`, {
        method: "POST",
        headers: {
          "content-type": "application/json",
          authorization: `Bearer ${src.apiKey}`,
        },
        body: JSON.stringify(body),
        signal: controller.signal,
      })
    } finally {
      clearTimeout(timer)
    }
  }

  /**
   * Judge ONE event and return its Verdict, or `null` when no runtime is
   * configured or the call failed (timeout, 429, 5xx, network). Deciding what
   * a missing verdict means is the CALLER's job — that is the deployment's
   * `fail_mode` (default OPEN, specification/degraded-mode.md), and the spec
   * is explicit that a 429 is an outage, not an allow.
   */
  async evaluate(event: WireEvent): Promise<WireVerdict | null> {
    try {
      // Stamped HERE, not at each construction site: one send path means the
      // build id cannot go missing on one kind of event only.
      const res = await this.post("/v1/evaluate", { ...event, integration: INTEGRATION })
      if (!res) return null
      if (!res.ok) {
        this.log.warn(`[openguardrails] evaluate answered ${res.status} — no verdict`)
        return null
      }
      return (await res.json()) as WireVerdict
    } catch (err) {
      this.log.warn(`[openguardrails] evaluate failed (${String(err)}) — no verdict`)
      return null
    }
  }

  /**
   * Integration liveness, fire-and-forget: a lost heartbeat is a lost signal,
   * never a lost enforcement, so it warns and moves on. This is where the
   * build id and the degraded-mode counters travel.
   */
  async heartbeat(body: WireHeartbeat): Promise<void> {
    try {
      const res = await this.post("/v1/heartbeat", body)
      if (res && !res.ok) {
        this.log.warn(`[openguardrails] heartbeat answered ${res.status}`)
      }
    } catch (err) {
      this.log.warn(`[openguardrails] heartbeat failed (${String(err)})`)
    }
  }
}

File: references/openguardrails-main/openguardrails-main/integrations/agent/litellm/openguardrails_litellm/wire.py
"""The OGR v0.8 wire, hand-rolled.

There is no SDK layer — the Runtime API is the integration surface
(specification/runtime-api.md), and an integration is one endpoint hit twice
per model call: ``POST /v1/evaluate`` holding the request, ``POST
/v1/evaluate`` holding the response. This module is that call plus the
optional heartbeat, on nothing but the standard library (``urllib.request``):
no batching machinery, no client-side decomposition, no third-party HTTP
dependency.

The base URL is joined with the canonical ``/v1/...`` paths exactly as the
binding requires — a deployment-specific prefix belongs IN the configured
base URL (``https://host/api/public/ogr``), never hard-coded here.
"""

from __future__ import annotations

import json
import logging
import urllib.error
import urllib.request

__version__ = "0.1.0"

logger = logging.getLogger("openguardrails")

#: WHO REPORTED IT — ``"name/version"``, on every event AND on the heartbeat.
#: One constant so the two can never name different builds.
INTEGRATION = f"openguardrails-litellm/{__version__}"

#: The nine REQUIRED v0.8 GuardEvent fields, exactly as
#: schema/guard-event.schema.json requires them (additionalProperties: false —
#: nothing outside these plus :data:`OPTIONAL_EVENT_FIELDS` may ride along).
EVENT_FIELDS = (
    "kind",
    "step_id",
    "agent_id",
    "agent_type",
    "agent_workspace",
    "agent_user",
    "llm_protocol",
    "payload",
)

#: The one OPTIONAL field (2026-08-17). It rode the heartbeat ALONE until then,
#: which could not answer "which build produced this traffic": a runtime keys its
#: liveness record on the integration NAME — it must, so a rollout updates that row
#: instead of minting a second and reporting the old build as dark — so every
#: replica overwrites the others' version.
OPTIONAL_EVENT_FIELDS = ("integration",)


def _json_default(value):
    """Last-resort serializer: litellm request dicts can carry datetimes and
    pydantic leftovers; the wire stringifies rather than dropping the event."""
    return str(value)


class Wire:
    """The one-endpoint client. Returns a Verdict dict, or ``None`` when the
    runtime could not answer (timeout, 429, 5xx, network) — deciding what a
    missing verdict means is the CALLER's job: that is the deployment's
    ``fail_mode``, and the degraded-mode spec is explicit that a 429 is an
    outage, not an allow."""

    def __init__(self, runtime_url: str, api_key: str, timeout: float = 5.0):
        self.base = (runtime_url or "").rstrip("/")
        self.api_key = api_key or ""
        self.timeout = timeout

    @property
    def configured(self) -> bool:
        """Whether a runtime is configured at all (URL and key both present)."""
        return bool(self.base and self.api_key)

    def _post(self, path: str, body: dict) -> "tuple[int, dict] | None":
        data = json.dumps(body, default=_json_default).encode("utf-8")
        request = urllib.request.Request(
            self.base + path,
            data=data,
            headers={
                "Content-Type": "application/json",
                "Authorization": "Bearer " + self.api_key,
            },
            method="POST",
        )
        try:
            with urllib.request.urlopen(request, timeout=self.timeout) as response:
                return response.status, json.loads(response.read().decode("utf-8"))
        except urllib.error.HTTPError as err:  # non-2xx WITH a response
            return err.code, {}
        except Exception as err:  # timeout, refused connection, bad JSON, DNS
            logger.warning("openguardrails: POST %s failed (%s)", path, err)
            return None

    def evaluate(self, event: dict) -> "dict | None":
        """Judge ONE GuardEvent; the whole protocol is this call. A non-2xx
        answer (429 included) is the same as no answer: no verdict."""
        # Stamped HERE, not at each construction site: one send path means the
        # build id cannot go missing on one kind of event only.
        answered = self._post("/v1/evaluate", {**event, "integration": INTEGRATION})
        if answered is None:
            return None
        status, verdict = answered
        if not 200 <= status < 300:
            logger.warning(
                "openguardrails: evaluate answered %d — no verdict", status
            )
            return None
        return verdict

    def heartbeat(self, body: dict) -> bool:
        """Integration liveness (transport-level, not a GuardEvent). Carries
        the build id and the counters that make a fail-open gap visible."""
        answered = self._post("/v1/heartbeat", body)
        return answered is not None and 200 <= answered[0] < 300
